File: android/app/src/main/java/com/shorthand/expander/Expander.kt
package com.shorthand.expander

import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

/** One shorthand: a trigger the user types and the text it expands to. */
data class Shorthand(val trigger: String, val expansion: String)

/** A trigger found at the caret; [start] is the index where the trigger begins. */
data class Match(val shorthand: Shorthand, val start: Int)

/** Expansion text with placeholders substituted and the caret offset inside it. */
data class Rendered(val text: String, val cursorOffset: Int)

/**
 * Pure matching and placeholder logic. Mirrors chrome-extension/src/expander.js
 * and the rules in shared/SPEC.md.
 */
object Expander {
    private val ESCAPED = Regex("""^\{\{([a-zA-Z]+)\}\}""")
    private val PLACEHOLDER = Regex("""^\{([a-zA-Z]+)\}""")

    /** A trigger may only fire after start-of-text, whitespace or punctuation. */
    fun isBoundary(ch: Char?): Boolean {
        if (ch == null) return true
        if (ch == '_' || ch.isLetter() || ch.isDigit()) return false
        val type = Character.getType(ch)
        return type != Character.LETTER_NUMBER.toInt() && type != Character.OTHER_NUMBER.toInt()
    }

    fun findMatch(textBeforeCaret: CharSequence?, shorthands: List<Shorthand>): Match? {
        if (textBeforeCaret.isNullOrEmpty() || shorthands.isEmpty()) return null
        val text = textBeforeCaret.toString()
        var best: Match? = null
        for (sh in shorthands) {
            val t = sh.trigger
            if (t.isEmpty() || t.length > text.length) continue
            if (!text.endsWith(t)) continue
            val start = text.length - t.length
            val before = if (start > 0) text[start - 1] else null
            if (!isBoundary(before)) continue
            if (best == null || t.length > best.shorthand.trigger.length) {
                best = Match(sh, start)
            }
        }
        return best
    }

    fun render(expansion: String?, now: Date = Date()): Rendered {
        val src = expansion ?: ""
        val date = SimpleDateFormat("yyyy-MM-dd", Locale.US).format(now)
        val time = SimpleDateFormat("HH:mm", Locale.US).format(now)
        val values = mapOf(
            "date" to date,
            "time" to time,
            "datetime" to "$date $time",
        )
        val sb = StringBuilder(src.length)
        var cursor = -1
        var i = 0
        while (i < src.length) {
            val ch = src[i]
            if (ch == '{') {
                val rest = src.substring(i)
                val esc = ESCAPED.find(rest)
                if (esc != null) {
                    sb.append('{').append(esc.groupValues[1]).append('}')
                    i += esc.value.length
                    continue
                }
                val m = PLACEHOLDER.find(rest)
                if (m != null) {
                    val name = m.groupValues[1].lowercase(Locale.ROOT)
                    if (name == "cursor") {
                        if (cursor < 0) cursor = sb.length
                        i += m.value.length
                        continue
                    }
                    val v = values[name]
                    if (v != null) {
                        sb.append(v)
                        i += m.value.length
                        continue
                    }
                }
            }
            sb.append(ch)
            i++
        }
        val text = sb.toString()
        return Rendered(text, if (cursor < 0) text.length else cursor)
    }

    /** Trim triggers, drop empties and duplicates (first one wins). */
    fun sanitize(list: List<Shorthand?>?): List<Shorthand> {
        if (list == null) return emptyList()
        val seen = HashSet<String>()
        val out = ArrayList<Shorthand>()
        for (item in list) {
            if (item == null) continue
            val trigger = item.trigger.trim()
            if (trigger.isEmpty() || !seen.add(trigger)) continue
            out.add(Shorthand(trigger, item.expansion))
        }
        return out
    }
}

File: windows/ShorthandExpander/Expander.cs
using System;
using System.Collections.Generic;
using System.Text;
using System.Text.RegularExpressions;

namespace ShorthandExpander
{
    public readonly struct Match
    {
        public Shorthand Shorthand { get; }
        /// <summary>Index in the text where the trigger starts.</summary>
        public int Start { get; }

        public Match(Shorthand shorthand, int start)
        {
            Shorthand = shorthand;
            Start = start;
        }
    }

    public readonly struct Rendered
    {
        public string Text { get; }
        /// <summary>Where the caret should end up, measured from the start of Text.</summary>
        public int CursorOffset { get; }

        public Rendered(string text, int cursorOffset)
        {
            Text = text;
            CursorOffset = cursorOffset;
        }
    }

    /// <summary>
    /// Pure matching and placeholder logic. Mirrors chrome-extension/src/expander.js
    /// and the rules in shared/SPEC.md.
    /// </summary>
    public static class Expander
    {
        private static readonly Regex Escaped = new Regex(@"^\{\{([a-zA-Z]+)\}\}", RegexOptions.Compiled);
        private static readonly Regex Placeholder = new Regex(@"^\{([a-zA-Z]+)\}", RegexOptions.Compiled);

        /// <summary>A trigger may only fire after start-of-text, whitespace or punctuation.</summary>
        public static bool IsBoundary(char? ch)
        {
            if (ch == null) return true;
            char c = ch.Value;
            return !(char.IsLetter(c) || char.IsNumber(c) || c == '_');
        }

        public static Match? FindMatch(string? textBeforeCaret, IReadOnlyList<Shorthand>? shorthands)
        {
            if (string.IsNullOrEmpty(textBeforeCaret) || shorthands == null || shorthands.Count == 0) return null;
            Match? best = null;
            foreach (var sh in shorthands)
            {
                string t = sh?.Trigger ?? "";
                if (t.Length == 0 || t.Length > textBeforeCaret.Length) continue;
                if (!textBeforeCaret.EndsWith(t, StringComparison.Ordinal)) continue;
                int start = textBeforeCaret.Length - t.Length;
                char? before = start > 0 ? textBeforeCaret[start - 1] : (char?)null;
                if (!IsBoundary(before)) continue;
                if (best == null || t.Length > best.Value.Shorthand.Trigger.Length)
                {
                    best = new Match(sh!, start);
                }
            }
            return best;
        }

        public static Rendered Render(string? expansion) => Render(expansion, DateTime.Now);

        public static Rendered Render(string? expansion, DateTime now)
        {
            string src = expansion ?? "";
            string date = now.ToString("yyyy-MM-dd");
            string time = now.ToString("HH:mm");
            var values = new Dictionary<string, string>(StringComparer.OrdinalIgnoreCase)
            {
                ["date"] = date,
                ["time"] = time,
                ["datetime"] = date + " " + time,
            };

            var sb = new StringBuilder(src.Length);
            int cursor = -1;
            int i = 0;
            while (i < src.Length)
            {
                char ch = src[i];
                if (ch == '{')
                {
                    string rest = src.Substring(i);
                    var esc = Escaped.Match(rest);
                    if (esc.Success)
                    {
                        sb.Append('{').Append(esc.Groups[1].Value).Append('}');
                        i += esc.Length;
                        continue;
                    }
                    var m = Placeholder.Match(rest);
                    if (m.Success)
                    {
                        string name = m.Groups[1].Value.ToLowerInvariant();
                        if (name == "cursor")
                        {
                            if (cursor < 0) cursor = sb.Length;
                            i += m.Length;
                            continue;
                        }
                        if (values.TryGetValue(name, out var v))
                        {
                            sb.Append(v);
                            i += m.Length;
                            continue;
                        }
                    }
                }
                sb.Append(ch);
                i++;
            }
            string text = sb.ToString();
            if (cursor < 0) cursor = text.Length;
            return new Rendered(text, cursor);
        }

        /// <summary>Trim, drop empties and duplicates (first one wins).</summary>
        public static List<Shorthand> Sanitize(IEnumerable<Shorthand?>? list)
        {
            var result = new List<Shorthand>();
            if (list == null) return result;
            var seen = new HashSet<string>(StringComparer.Ordinal);
            foreach (var item in list)
            {
                if (item == null) continue;
                string trigger = (item.Trigger ?? "").Trim();
                if (trigger.Length == 0 || !seen.Add(trigger)) continue;
                result.Add(new Shorthand(trigger, item.Expansion ?? ""));
            }
            return result;
        }
    }
}
